cmd/backup-manager/app/compact/manager.go

```diff
@@ -0,0 +1,229 @@
+	return string(out), nil
+}
+
+func (cm *Manager) base64ifyCmd(ctx context.Context) (*exec.Cmd, error) {
+	br := cm.brBin()
+	args := []string{
+		"operator",
+		"base64ify",
+	}
+	StorageOpts, err := pkgutil.GenStorageArgsForFlag(cm.compact.Spec.StorageProvider, "storage")
+	if err != nil {
+		return nil, err
+	}
+	args = append(args, StorageOpts...)
+	return exec.CommandContext(ctx, br, args...), nil
+}
+
+func (cm *Manager) runCompaction(ctx context.Context, base64Storage string) (err error) {
+	cmd := cm.compactCmd(ctx, base64Storage)
+
+	logs, err := cmd.StderrPipe()
+	if err != nil {
+		return errors.Annotate(err, "failed to create stderr pipe for compact")
+	}
+	if err := cmd.Start(); err != nil {
+		return errors.Annotate(err, "failed to start compact")
+	}
+
+	cm.statusUpdater.OnStart(ctx, cm.compact)
+	err = cm.processCompactionLogs(ctx, io.TeeReader(logs, os.Stdout))
+	if err != nil {
+		return err
+	}
+
+	return cmd.Wait()
+}
+
+func (cm *Manager) compactCmd(ctx context.Context, base64Storage string) *exec.Cmd {
+	ctl := cm.kvCtlBin()
+	// You should not change the log configuration here, it should sync with the upstream setting
+	args := []string{
+		"--log-level",
+		"INFO",
+		"--log-format",
+		"json",
+		"compact-log-backup",
+		"--storage-base64",
+		base64Storage,
+		"--from",
+		strconv.FormatUint(cm.options.FromTS, 10),
+		"--until",
+		strconv.FormatUint(cm.options.UntilTS, 10),
+		"-N",
+		strconv.FormatUint(cm.options.Concurrency, 10),
+	}
+	return exec.CommandContext(ctx, ctl, args...)
+}
+
+func (cm *Manager) processCompactionLogs(ctx context.Context, logStream io.Reader) error {
+	dec := json.NewDecoder(logStream)
+
+	for dec.More() {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		var raw json.RawMessage
+		if err := dec.Decode(&raw); err != nil {
+			return errors.Annotate(err, "failed to decode raw log line")
+		}
+
+		var line logLine
+		if err := json.Unmarshal(raw, &line); err != nil {
+			return errors.Annotate(err, "failed to decode the line of log")
+		}
+		line.Raw = raw
+
+		if err := cm.processLogLine(ctx, line); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (cm *Manager) processLogLine(ctx context.Context, l logLine) error {
+	const (
+		messageCompactionDone = "Finishing compaction."
+		messageCompactAborted = "Compaction aborted."
+	)
+
+	switch l.Message {
+	case messageCompactionDone:
+		var prog Progress
+		if err := json.Unmarshal(l.Raw, &prog); err != nil {
+			return errors.Annotatef(err, "failed to decode progress message: %s", string(l.Raw))
+		}
+		cm.statusUpdater.OnProgress(ctx, prog, cm.compact)
+		return nil
+	case messageCompactAborted:
+		errContainer := struct {
+			Err string `json:"err"`
+		}{}
+		if err := json.Unmarshal(l.Raw, &errContainer); err != nil {
+			return errors.Annotatef(err, "failed to decode error message: %s", string(l.Raw))
+		}
+		return errors.New(errContainer.Err)
+	default:
+		klog.Infof("progress log: %s", l.Message)
```

Comment:
Since you already use the `io.TeeReader` to forward log to stdout, you don't need to log message again.